Feature/bjs2 77981 skills

tanufa25 opened the pull request:
Создание и настройка скиллов

Commit:
BJS2-77981
Создание Skill

Commit:
BJS2-77981
Получение списка скиллов пользователя

Commit:
BJS2-77981
Получение списка предложенных скиллов пользователя.
Поправила получение скиллов: теперь вывод для текущего пользователя

Commit:
BJS2-77981
UserSkillController создан для получения навыков пользователя
SkillMapper переписан без mapstruct

Commit:
BJS2-77981
Приобретение предложенного скилла

Commit:
BJS2-77981
Использован существующий класс констант

FROLOV-KI commented on src/main/java/school/faang/user_service/util/LogsConstants.java:
```diff
@@ -18,4 +18,16 @@ public final class LogsConstants {
+    public static final String USER_HAS_SKILL = "У пользователя %s уже есть навык %s";
+    public static final String SKILL_NOT_FOUND = "Навык по id = %s не найден";
+
+    public static final int MIN_SKILL_OFFERS = 3;
```
Мб вынеси в свойста приложения, чтобы можно было гибко настраивать?

tanufa25 replied:
@20cognosce, подскажи, пожалуйста, в константах этих оставить или отдельный файл для констант-настроек или всё же, как Кирилл говорит, в application.yaml, на сколько я поняла.

20cognosce replied:
В application.yaml не надо это выносить, оставьте в коде
можно в отдельный класс чисто под константы для ошибок

FROLOV-KI commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,87 @@
+            User user = userRepository.findById(userId).
+                    orElseThrow(() -> new EntityNotFoundException(String.format(USER_NOT_FOUND, userId)));
+            skillRepository.assignSkillToUser(skillId, userId);
+            skillOfferList.forEach((skillOffer) -> {
```
Мб вынести эту логику в отдельный приватный метод. А то выходит большой метод с большим количеством логики, тежко воспринимается

FROLOV-KI commented on src/main/java/school/faang/user_service/mapper/SkillMapper.java:
```diff
@@ -0,0 +1,65 @@
+public class SkillMapper {
+
+    public Skill toSkill(CreateSkillDto dto) {
+        if (dto == null) {
```
мб как-то проверять перед маппером, что не null? Словно это потенциальная проблема с NPE

tanufa25 replied:
Я посмотрела, что в mapStruct под капотом так и работает, поэтому так и написала.
@20cognosce, норм так? Или всё же сделать, как Кирилл говорит?

20cognosce replied:
перед маппером не надо, в маппере кидай ошибку

FROLOV-KI commented on src/main/java/school/faang/user_service/mapper/SkillMapper.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    public List<SkillDto> toSkillDtos(List<Skill> skills) {
+        if (skills == null) {
```
+

FROLOV-KI commented on src/main/java/school/faang/user_service/mapper/SkillMapper.java:
```diff
@@ -0,0 +1,65 @@
+
+    public SkillDto toDto(Skill skill) {
+        if (skill == null) {
+            return null;
```
+

FROLOV-KI commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,87 @@
+            });
+
+        } else {
+            skill = null;
```
Не опасно ли возвращать null?

Fagan89 commented on src/main/java/school/faang/user_service/exception/EntityNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package school.faang.user_service.exception;
+
+public class EntityNotFoundException extends RuntimeException  {
```
такое исключение есть в джава, не надо создавать свое
![2025-05-16_00-38-48](https://github.com/user-attachments/assets/25f03f42-875b-481d-b5f4-568e5e202c5f)

tanufa25 replied:
Удалила свой класс, использовала вместо него чей-то RecordNotFoundException

Commit:
BJS2-77981
Optional добавлен в UserService в методе acquireSkillFromOffers.
В UserSkillController добавлен статус ответа PRECONDITION_FAILED

20cognosce commented on src/main/java/school/faang/user_service/mapper/SkillMapper.java:
```diff
@@ -0,0 +1,65 @@
+
+    public Skill toSkill(CreateSkillDto dto) {
+        if (dto == null) {
+            return null;
```
вот так точно делать не надо, кидай ошибку

20cognosce commented on src/main/java/school/faang/user_service/mapper/SkillMapper.java:
```diff
@@ -0,0 +1,65 @@
+
+    public List<SkillDto> toSkillDtos(List<Skill> skills) {
+        if (skills == null) {
+            return null;
```
никогда не возвращать null

20cognosce commented on src/main/java/school/faang/user_service/entity/User.java:
```diff
@@ -40,6 +42,11 @@
 @Builder
 @NoArgsConstructor
 @AllArgsConstructor
+@ToString(exclude = {"country", "followers", "followees", "ownedEvents", "mentees", "mentors",
+        "receivedMentorshipRequests", "sentMentorshipRequests", "sentGoalInvitations",
+        "receivedGoalInvitations", "setGoals", "goals", "skills", "participatedEvents",
+        "recommendationsGiven", "recommendationsReceived", "contacts", "ratings",
+        "userProfilePic", "contactPreference", "premium", "education", "career", "workSchedule"})
```
чет слишком жестко
только @OneToMany исключай, либо где циклические зависимости, зачем всё...

tanufa25 replied:
А почему только @onetomany, а @manytomany?
И в моём понимании как раз все вышеперечисленные цикличные.

20cognosce commented on src/main/java/school/faang/user_service/dto/skill/SkillDto.java:
```diff
@@ -0,0 +1,15 @@
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
```
посмотри что она в себя включает и убери лишнее

20cognosce commented on src/main/java/school/faang/user_service/controller/skill/SkillController.java:
```diff
@@ -0,0 +1,45 @@
+            return ResponseEntity
+                    .status(HttpStatus.CREATED)
+                    .body(skillDtoCreated);
+        } catch (DataValidationException e) {
+            return ResponseEntity
+                    .status(HttpStatus.BAD_REQUEST)
+                    .body(Collections.singletonMap("error", e.getMessage()));
+        }
```
вот этого вот в контроллере вообще быть не должно

если кидается ошибка, этот маппинг в @ExceptionHandler делается

20cognosce commented on src/main/java/school/faang/user_service/controller/skill/SkillController.java:
```diff
@@ -0,0 +1,45 @@
+            Skill skill = skillMapper.toSkill(skillDto);
+            skill = skillService.create(skill);
+            SkillDto skillDtoCreated = skillMapper.toDto(skill);
+            log.info("Созданный навык: {}", skillDtoCreated);
```
в сервисе все логи

20cognosce commented on src/main/java/school/faang/user_service/dto/skill/CreateSkillDto.java:
```diff
@@ -0,0 +1,12 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class CreateSkillDto {
```
Делай однообразный нейминг SkillDto SkillCreateDto SkillUpdateDto SkillCandidateDto

20cognosce commented on src/main/java/school/faang/user_service/dto/skill/CreateSkillDto.java:
```diff
@@ -0,0 +1,12 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class CreateSkillDto {
+    private String title;
```
валидация входных параметров где у нас?

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,90 @@
+
+    @Transactional
+    public Skill create(Skill skill) {
+        skillValidator.validateTitleBlank(skill.getTitle());
```
вот это аннотациями делается, это проверка на дурака - в ней не нужна никакая логика, чтобы это выносить в валидатор

20cognosce commented on src/main/java/school/faang/user_service/controller/skill/UserSkillController.java:
```diff
@@ -0,0 +1,71 @@
+    @GetMapping("/{userId}/skills/offered")
+    public ResponseEntity<?> getOfferedSkills(@PathVariable long userId) {
+        List<SkillCandidateDto> offeredSkillDtoList = skillMapper.toOfferedDtos(skillService.getOfferedSkills(userId));
+        log.info("У пользователя {} предложенные навыки {}", userId, offeredSkillDtoList);
```
не надо в контроллере писать логи

20cognosce replied:
и тем более на кириллице

20cognosce replied:
тем более какой смысл этого лога, если есть ответ успешный - мы итак увидим, что пришло

20cognosce commented on src/main/java/school/faang/user_service/controller/skill/UserSkillController.java:
```diff
@@ -0,0 +1,71 @@
+
+    @GetMapping("/{userId}/skills/offered")
+    public ResponseEntity<?> getOfferedSkills(@PathVariable long userId) {
+        List<SkillCandidateDto> offeredSkillDtoList = skillMapper.toOfferedDtos(skillService.getOfferedSkills(userId));
```
в одну строчку писать не надо, создавай переменные

20cognosce commented on src/main/java/school/faang/user_service/controller/skill/UserSkillController.java:
```diff
@@ -0,0 +1,71 @@
+
+    @PostMapping("/{userId}/skills/offered/{skillId}")
+    public ResponseEntity<?> acquireSkillFromOffers(@PathVariable long userId, @PathVariable long skillId) {
+        try {
+            Optional<Skill> skillOpt = skillService.acquireSkillFromOffers(userId, skillId);
+            if (skillOpt.isPresent()) {
+                SkillDto skillDto = skillMapper.toDto(skillOpt.get());
+                log.info("Пользователю {} присвоен навык {}", userId, skillDto);
+                return ResponseEntity
+                        .status(HttpStatus.OK)
+                        .body(skillDto);
+            } else {
+                log.warn(CONDITION_FOR_OFFERS_AMOUNT_FAILED);
+                return ResponseEntity
+                        .status(HttpStatus.PRECONDITION_FAILED)
+                        .body(Collections.singletonMap("warning", CONDITION_FOR_OFFERS_AMOUNT_FAILED));
+            }
+        } catch (DataValidationException | EntityNotFoundException e) {
+            return ResponseEntity
+                    .status(HttpStatus.BAD_REQUEST)
+                    .body(Collections.singletonMap("error", e.getMessage()));
+        }
```
переделать это всё, в контроллере каждый метод 3-4 +- строчки должен занимать, какие условия, какие try catch
no no no

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,90 @@
+            skillRepository.assignSkillToUser(skillId, userId);
+            skillOfferList.forEach((skillOffer) -> fillUserSkillGuarantee(skillOffer, user, skill));
+            return Optional.ofNullable(skill);
+        } else {
+            return Optional.empty();
+        }
```
а можно развернуть условие и сразу пустой опшинал отдать, чтобы вложенность не плодить

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,90 @@
+
+    private void fillUserSkillGuarantee(SkillOffer skillOffer, User user, Skill skill) {
+        Recommendation recommendation = skillOffer.getRecommendation();
+        if (recommendation != null) {
```
непрозрачное поведение, ошибку кидай явно, если что-то не так

Commit:
BJS2-77981
GlobalExceptionHandler добавлен.
UserSkillController исправлен.
Язык исключений и логов - английский.
Exceptions в skillMapper.

Commit:
BJS2-77981
EntityNotFoundException удалён, вместо него используется RecordNotFoundException
SkillValidator перемещён

20cognosce commented on src/main/java/school/faang/user_service/exception/config/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,96 @@
+@RestControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler({
+            DataValidationException.class,
+            EventRegisterException.class
+    })
+    public ResponseEntity<ErrorResponse> handleBadRequestExceptions(RuntimeException e) {
+        ErrorResponse error = new ErrorResponse(
+                HttpStatus.BAD_REQUEST.value(),
+                e.getMessage()
+        );
+        return new ResponseEntity<>(error, HttpStatus.BAD_REQUEST);
```
в зависимости от ошибки меняется только статус код в 90% случаев.
можно сделать мапу: класс ошибки - статус код

20cognosce replied:
И сделать единый handler

20cognosce commented on src/main/java/school/faang/user_service/service/skill/SkillService.java:
```diff
@@ -0,0 +1,95 @@
+        log.info("Create skill: {}", skill);
+        skillValidator.validateTitleUnique(skill.getTitle());
+        Skill createdSkill = skillRepository.save(skill);
+        log.info("Created Skill: {}", createdSkill);
```
ты пишешь слишком много одинаковых логов

20cognosce commented on src/main/java/school/faang/user_service/service/skill/SkillService.java:
```diff
@@ -0,0 +1,95 @@
+
+    @Transactional
+    public Skill acquireSkillFromOffers(long userId, long skillId) {
+        log.info("acquireSkillFromOffers userId = {}, skillId = {}", userId, skillId);
+        Skill skill = skillRepository.findById(skillId).
+                orElseThrow(() -> new RecordNotFoundException(String.format(SKILL_NOT_FOUND, skillId)));
+        User user = userRepository.findById(userId).
+                orElseThrow(() -> new RecordNotFoundException(String.format(USER_NOT_FOUND, userId)));
+        skillValidator.validateUserHasSkill(userId, skillId);
+        List<SkillOffer> skillOfferList = skillOfferRepository.findAllOffersOfSkill(skillId, userId);
+        log.info("acquireSkillFromOffers skillOfferList = {}", skillOfferList);
+        int offeredAmount = skillOfferList.size();
+        if (offeredAmount < MIN_SKILL_OFFERS) {
+            log.error(CONDITION_FOR_OFFERS_AMOUNT_FAILED);
+            throw new PreConditionFailedException(CONDITION_FOR_OFFERS_AMOUNT_FAILED);
+        }
+        skillRepository.assignSkillToUser(skillId, userId);
+        skillOfferList.forEach((skillOffer) -> fillUserSkillGuarantee(skillOffer, user, skill));
+        log.info("acquireSkillFromOffers skill = {}", skill);
+        return skill;
```
как-нибудь декомпозировать и уменьшить метод процентов на 30 можно
